File: rust/tauri-app/src-tauri/src/wallet_list_fs.rs
use serde_json::json;
use std::fs;
use std::path::Path;

/// Same as `listWallets` in `wallet-rpc.js` (file scan; legacy dir omitted in first version).
pub fn list_wallet_files (wallet_dir: &Path) -> Result<serde_json::Value, String> {
  let mut list = vec![];
  let mut directories = vec![];
  if !wallet_dir.is_dir() {
    return Ok(json!({ "list": [], "directories": [] }));
  }
  for ent in fs::read_dir(wallet_dir).map_err(|e| e.to_string())? {
    let ent = ent.map_err(|e| e.to_string())?;
    let filename = ent.file_name();
    let name = filename.to_string_lossy().to_string();
    match name.as_str() {
      ".DS_Store" | ".DS_Store?" | "._.DS_Store" | ".Spotlight-V100" | ".Trashes" | "Thumbs.db" | "ehthumbs.db" | "old-gui" => continue,
      _ => {}
    }
    let p = ent.path();
    if p.is_dir() {
      let wfile = p.join(&name);
      let keyf = wfile.to_string_lossy().to_string() + ".keys";
      if wfile.is_file() && Path::new(&keyf).is_file() {
        directories.push(name);
      }
      continue;
    }
    if p
      .extension()
      .is_some() {
      continue;
    }
    let mut wallet_data = json!({
      "name": name,
      "address": serde_json::Value::Null,
      "password_protected": serde_json::Value::Null
    });
    let meta = wallet_dir.join(format!("{name}.meta.json"));
    if meta.is_file() {
      if let Ok(s) = fs::read_to_string(&meta) {
        if let Ok(m) = serde_json::from_str::<serde_json::Value>(&s) {
          if let Some(a) = m.get("address") {
            wallet_data["address"] = a.clone();
          }
          if let Some(p) = m.get("password_protected") {
            wallet_data["password_protected"] = p.clone();
          }
        }
      }
    }
    let addrf = wallet_dir.join(format!("{name}.address.txt"));
    if addrf.is_file() {
      if let Ok(s) = fs::read_to_string(&addrf) {
        if !s.is_empty() {
          wallet_data["address"] = json!(s.trim());
        }
      }
    }
    list.push(wallet_data);
  }
  Ok(json!({ "list": list, "directories": directories }))
}

File: flutter/arqma_wallet_gui/lib/core/desktop/wallet_list_fs.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/foundation.dart';

/// Same rules as `wallet_list_fs::list_wallet_files` (`rust/tauri-app/src-tauri/src/wallet_list_fs.rs`).
Map<String, dynamic> listWalletFiles(String walletDirPath) {
  final Directory dir = Directory(walletDirPath);
  if (!dir.existsSync()) {
    return <String, dynamic>{'list': <dynamic>[], 'directories': <dynamic>[], 'legacy': <dynamic>[]};
  }
  final List<Map<String, dynamic>> list = <Map<String, dynamic>>[];
  final List<String> directories = <String>[];
  const Set<String> skipNames = <String>{
    '.DS_Store',
    '.DS_Store?',
    '._.DS_Store',
    '.Spotlight-V100',
    '.Trashes',
    'Thumbs.db',
    'ehthumbs.db',
    'old-gui',
  };
  try {
    for (final FileSystemEntity ent in dir.listSync(followLinks: false)) {
      final String name = ent.path.split(Platform.pathSeparator).last;
      if (skipNames.contains(name)) {
        continue;
      }
      if (ent is Directory) {
        final File wfile = File('${ent.path}${Platform.pathSeparator}$name');
        final File keyf = File('${wfile.path}.keys');
        if (wfile.existsSync() && keyf.existsSync()) {
          directories.add(name);
        }
        continue;
      }
      if (ent is! File) {
        continue;
      }
      if (name.contains('.')) {
        continue;
      }
      final Map<String, dynamic> walletData = <String, dynamic>{
        'name': name,
        'address': null,
        'password_protected': null,
      };
      final File meta = File('${dir.path}${Platform.pathSeparator}$name.meta.json');
      if (meta.existsSync()) {
        try {
          final dynamic m = jsonDecode(meta.readAsStringSync());
          if (m is Map) {
            final Map<String, dynamic> mm = Map<String, dynamic>.from(m);
            if (mm.containsKey('address')) {
              walletData['address'] = mm['address'];
            }
            if (mm.containsKey('password_protected')) {
              walletData['password_protected'] = mm['password_protected'];
            }
          }
        } catch (e) {
          debugPrint('[wallet_list_fs] meta $name: $e');
        }
      }
      final File addrf = File('${dir.path}${Platform.pathSeparator}$name.address.txt');
      if (addrf.existsSync()) {
        final String s = addrf.readAsStringSync().trim();
        if (s.isNotEmpty) {
          walletData['address'] = s;
        }
      }
      list.add(walletData);
    }
  } catch (e, st) {
    debugPrint('[wallet_list_fs] $e\n$st');
    return <String, dynamic>{'list': <dynamic>[], 'directories': <dynamic>[], 'legacy': <dynamic>[]};
  }
  return <String, dynamic>{'list': list, 'directories': directories, 'legacy': <dynamic>[]};
}
